posts/2021-04-30-edition-2021.md

```diff
@@ -0,0 +1,346 @@
+automatically capture anything that you refer to from within their body.
+For example, `|| a + 1` automatically captures a reference to `a` from the surrounding context.
+
+Currently, this applies to whole structs, even when only using one field.
+For example, `|| a.x + 1` captures a reference to `a` and not just `a.x`.
+In some situations, this is a problem.
+When a field of the struct is already borrowed (mutably) or moved out of,
+the other fields can no longer be used in a closure,
+since that would capture the whole struct, which is no longer available.
+
+```rust
+let a = SomeStruct::new();
+
+drop(a.x); // Move out of one field of the struct
+
+println!("{}", a.y); // Ok: Still use another field of the struct
+
+let c = || println!("{}", a.y); // Error: Tries to capture all of `a`
+c();
+```
+
+Starting in Rust 2021, closures will only capture the fields that they use.
+So, the above example will compile fine in Rust 2021.
+
+This new behavior is only activated in the new edition,
+since it can change the order in which fields are dropped.
+As for all edition changes, an automatic migration is available,
+which will update your closures for which this matters.
+It can insert `let _ = &a;` inside the closure to force the entire
+struct to be captured as before.
+
+### Panic macro consistency
+
+The `panic!()` macro is one of Rust's most well known macros.
+However, it has [some subtle surprises](https://github.com/rust-lang/rfcs/blob/master/text/3007-panic-plan.md)
+that we can't just change due to backwards compatibility.
+
+```rust
+panic!("{}", 1); // Ok, panics with the message "1"
+panic!("{}"); // Ok, panics with the message "{}"
+```
+
+The `panic!()` macro only uses string formatting when it's invoked with more than one argument.
+When invoked with a single argument, it doesn't even look at that argument.
+
+```rust
+let a = "{";
+println!(a); // Error: First argument must be a format string literal
+panic!(a); // Ok: The panic macro doesn't care
+```
+
+(It even accepts non-strings such as `panic!(123)`, which is uncommon and rarely useful.)
+
+This is especially a problem when
+[implicit format arguments](https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html)
+will be stabilized.
+That feature will make `println!("hello {name}")` a short-hand for `println!("hello {}", name)`.
+However, `panic!("hello {name}")` would not work as expected,
+since `panic!()` doesn't process a single argument as format string.
+
+To avoid that confusing situation, Rust 2021 features a more consistent `panic!()` macro.
+The new `panic!()` macro will no longer accept arbitrary expressions as the only argument.
+It will&mdash;just like `println!()`&mdash;always process the first argument as format string.
+
+In addition, `core::panic!()` and `std::panic!()` will be identical in Rust 2021.
+Currently, there are some historical differences between those two,
+which can be noticable when switching `#![no_std]` on or off.
```

Comment:
I think a brief reference to [panic_any](https://doc.rust-lang.org/nightly/std/panic/fn.panic_any.html) here seems like a good idea.